src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs

```diff
@@ -267,59 +346,57 @@ public override BoundNode VisitFixedLocalCollectionInitializer(BoundFixedLocalCo
             BoundExpression addition = factory.Binary(BinaryOperatorKind.PointerAndIntAddition, localType, factory.Local(localSymbol), helperCall);
             BoundStatement conditionalAdd = factory.If(notNullCheck, factory.Assignment(factory.Local(localSymbol), addition));
 
-            localToClear = stringTemp;
             return factory.Block(stringTempInit, localInit, conditionalAdd);
         }
 
+        // fixed(int* ptr = arr){ ... }    == becomes ===>
+        // 
+        // pinned int[] pinnedTemp = arr;         // pinning managed ref
+        // int* ptr = pinnedTemp != null && pinnedTemp.Length != 0
+        //                (int*)&pinnedTemp[0]:   // unsafe cast to unmanaged ptr
+        //                0;
+        //   . . . 
         private BoundStatement InitializeFixedStatementArrayLocal(
             BoundLocalDeclaration localDecl,
             LocalSymbol localSymbol,
             BoundFixedLocalCollectionInitializer fixedInitializer,
             SyntheticBoundNodeFactory factory,
-            out LocalSymbol arrayTemp)
+            out LocalSymbol pinnedTemp)
         {
-            // From ExpressionBinder::BindPtrToArray:
-            // (((temp = array) != null && temp.Length > 0) ? loc = &temp[0] : loc = null)
-            // NOTE: The assignment needs to be inside the EK_QUESTIONMARK. See Whidbey bug #397859.
-            // We can't do loc = (... ? ... : ...) since the CLR type of &temp[0] is a managed
-            // pointer and null is a UIntPtr - which confuses the JIT. We can't just convert
-            // &temp[0] to UIntPtr with a conv.u instruction because then if a GC occurs between
-            // the time of the cast and the assignment to the local, we're toast.
```

Comment:
This subtlety is a another consequence of conflating an unmanaged pointer and a managed reference in one variable.